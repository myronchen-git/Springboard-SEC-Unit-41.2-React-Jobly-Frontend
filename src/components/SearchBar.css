.SearchBar button {
  width: 100%;
}
