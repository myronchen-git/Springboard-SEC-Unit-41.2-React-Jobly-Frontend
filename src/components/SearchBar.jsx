import PropTypes from 'prop-types';
import { useState } from 'react';
import { Button, Col, Container, Form, Input, Row } from 'reactstrap';

import './SearchBar.css';

// ==================================================

/**
 * Provides a search bar for searching the list of all items.  Only allows
 * searching company names, job titles, or similar.
 *
 * @param {Object} props - React component properties.
 * @param {Function} props.retrieveItems - Calls Jobly's backend API to
 *   retrieve all items, using a given filter.
 * @param {String} props.filterName - Name of the filter to be used in the API
 *   call.
 */
function SearchBar({ retrieveItems, filterName }) {
  const initialFormData = { [filterName]: '' };
  const [formData, setFormData] = useState(initialFormData);

  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData((formData) => ({ ...formData, [name]: value }));
  }

  async function handleSubmit(evt) {
    evt.preventDefault();
    const filtersToSend = formData[filterName] === '' ? {} : formData;
    await retrieveItems(filtersToSend);
  }

  return (
    <Form className="SearchBar" onSubmit={handleSubmit}>
      <Container className="p-0" fluid>
        <Row className="g-3" xs="1" sm="2">
          <Col sm="9">
            <Input
              type="search"
              name={filterName}
              placeholder="Enter search term..."
              value={formData[filterName]}
              onChange={handleChange}
            />
          </Col>
          <Col sm="3">
            <Button type="submit">Search</Button>
          </Col>
        </Row>
      </Container>
    </Form>
  );
}

// ==================================================

SearchBar.propTypes = {
  retrieveItems: PropTypes.func.isRequired,
  filterName: PropTypes.string.isRequired,
};

export default SearchBar;
