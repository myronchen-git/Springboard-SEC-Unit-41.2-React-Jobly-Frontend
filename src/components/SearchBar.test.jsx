import { fireEvent, render } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';

import SearchBar from './SearchBar.jsx';

// ==================================================

describe('CompanyCard', () => {
  const mockRetrieveItems = vi.fn();

  beforeEach(() => {
    mockRetrieveItems.mockReset();
  });

  it('renders.', () => {
    render(<SearchBar retrieveItems={mockRetrieveItems} filterName="name" />);
  });

  it('matches snapshot.', () => {
    const { asFragment } = render(
      <SearchBar retrieveItems={mockRetrieveItems} filterName="name" />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('can update form data.', () => {
    // Arrange
    const { getByPlaceholderText } = render(
      <SearchBar retrieveItems={mockRetrieveItems} filterName="name" />
    );

    const searchInput = getByPlaceholderText('Enter search term...');

    // Act
    fireEvent.change(searchInput, { target: { value: 'abc' } });

    // Assert
    expect(searchInput).toHaveValue('abc');
    expect(mockRetrieveItems).not.toHaveBeenCalled();
  });

  it.each([
    ['abc', { name: 'abc' }],
    ['', {}],
  ])('can submit form data of "%s".', (name, expectedFilters) => {
    // Arrange
    const { getByPlaceholderText, getByText } = render(
      <SearchBar retrieveItems={mockRetrieveItems} filterName="name" />
    );

    const searchInput = getByPlaceholderText('Enter search term...');
    const submitBtn = getByText('Search');

    fireEvent.change(searchInput, { target: { value: name } });

    // Act
    fireEvent.click(submitBtn);

    // Assert
    expect(mockRetrieveItems).toHaveBeenCalledOnce();
    expect(mockRetrieveItems).toHaveBeenCalledWith(expectedFilters);
  });
});
