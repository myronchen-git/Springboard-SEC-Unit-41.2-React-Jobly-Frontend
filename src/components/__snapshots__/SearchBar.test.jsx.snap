// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CompanyCard > matches snapshot. 1`] = `
<DocumentFragment>
  bound HTMLFormElement {
    "0": <input
      class="form-control"
      name="name"
      placeholder="Enter search term..."
      type="search"
      value=""
    />,
    "1": <button
      class="btn btn-secondary"
      type="submit"
    >
      Search
    </button>,
  }
</DocumentFragment>
`;
